src/audio/dcblock/dcblock.c

```diff
@@ -0,0 +1,434 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Google LLC. All rights reserved.
+//
+// Author: Sebastiano Carlucci <scarlucci@google.com>
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/dcblock/dcblock.h>
+#include <sof/common.h>
+#include <sof/debug/panic.h>
+#include <sof/drivers/ipc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <sof/string.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/**
+ * \brief Sets the DC Blocking filter in pass through mode.
+ * The frequency response of a DCB filter is:
+ * H(z) = (1 - z^-1)/(1-Rz^-1).
+ * Setting R to 1 makes the filter act as a passthrough component.
+ */
+static int dcblock_set_passthrough(struct comp_data *cd)
+{
+	comp_cl_info(&comp_dcblock, "dcblock_set_passthrough()");
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		cd->R_coeffs[i] = ONE_Q2_30;
+
+	return 0;
+}
+
+/**
+ * \brief Initializes the state of the DC Blocking Filter
+ */
+static int dcblock_init_state(struct comp_data *cd,
+			      struct audio_stream *src_stream)
+{
+	int32_t *x0;
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++) {
+		x0 = audio_stream_read_frag_s32(src_stream, 0);
```

Comment:
I'm still wondering what's the purpose of initializing filter delay with first input sample.  That same sample will be fed again to filter when copy() starts.

Since it's in the non-recursive side the impact is minimal and not much worry. Does it give advantage vs. starting with just all zeroed delay line and less code?

Reply:
It helps in reducing the DC component from the first sample.
y[n] = x[n] - x[n-1] + yR[n-1]

If x[-1] is initialized to x[0], then y[0] is 0, filtering out the DC offset from the beginning. 

Is this behavior unnecessary since the volume component already ramps up the signal?

Reply:
Since the volume control is better to be after DC blocker it will be there to ramp the beginning, yes. I'd say this feature is not necessary but prefer to leave the decision to you.

Reply:
@singalsu I think we might want to use this as I am thinking we might tune the ramp in the volume control a bit differently. Currently we use about 20ms on the audio server where as the ramp SOF uses is quite noticeable.

Reply:
Also the pipeline and DAI is started after prepare() with trigger start. There may not be samples in the buffer. Have you tried if that works without xrun?

Reply:
You're right. The buffers did not have any samples at the prepare() function. 

Anyway, the initialization function has been updated to set x_prev to 0. The ramp should be enough to help smooth down the offset.

Reply:
Yep, that will be OK. If you'd like to experiment you could in copy() function consume the first sample twice and discard first output to get the same result.